sql/mysqld.cc

```diff
@@ -2804,14 +2805,20 @@ static void network_init(void)
     bzero((char*) &UNIXaddr, sizeof(UNIXaddr));
     UNIXaddr.sun_family = AF_UNIX;
     strmov(UNIXaddr.sun_path, mysqld_unix_port);
-    (void) unlink(mysqld_unix_port);
+#if defined(__linux__)
+    /* Abstract socket */
+    if (mysqld_unix_port[0] == '@')
+      UNIXaddr.sun_path[0]= '\0';
+    else
+#endif
+      (void) unlink(mysqld_unix_port);
     arg= 1;
     (void) mysql_socket_setsockopt(unix_sock,SOL_SOCKET,SO_REUSEADDR,
                                    (char*)&arg, sizeof(arg));
     umask(0);
     if (mysql_socket_bind(unix_sock,
                           reinterpret_cast<struct sockaddr *>(&UNIXaddr),
-                          sizeof(UNIXaddr)) < 0)
+                          sizeof(UNIXaddr.sun_family) + port_len) < 0)
```

Comment:
On OS X `sockaddr_un` is defined as:
```
struct sockaddr_un
{
  u_char sun_len;
  u_char sun_family;
  char  sun_path[104];
};
```

When I suggested `sizeof(UNIXaddr.sun_family) + port_len` I thought it would be undef `#if defined(__linux__)`.

There is a comment in man 7 unix (on linux):
> On  Linux, the above offsetof() expression equates to the same value as sizeof(sa_family_t), but some other implementations include other fields before sun_path, so the offsetof() expression more portably describes the size of the address structure.

So please either use `offsetof()` here or move calculation above around `__linux__` and `@`-sign checks.